### PR TITLE
feature: image & app hub

### PR DESCRIPTION
**Pull request by lingdie**
issue link: https://github.com/labring/sealos/issues/1902#issue-1411314459

### CODE REVIEW

**Comment by fanux** [comment]
`git commit -s --amend && git push -f` to signoff your commit.

File: controllers/app/main.go [hunk 1]
```diff
@@ -71,7 +67,7 @@ func main() {
 		Port:                   9443,
 		HealthProbeBindAddress: probeAddr,
 		LeaderElection:         enableLeaderElection,
-		LeaderElectionID:       "5d8f3b44.sealos.io",
+		LeaderElectionID:       "f3e714c5.sealos.io",
```
Comment (fanux) [inline controllers/app/main.go:70]: I Think we can generate controller code for app, to prevent future need to use

**Comment by lingdie** [comment]
draft pr, I will fix these problems later.

File: controllers/image/api/v1/image_types.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// ImageSpec defines the desired state of Image
+type ImageSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+	Name string `json:"name"`
+	Icon string `json:"icon,omitempty"`
+	URL  string `json:"url,omitempty"`
+	Type string `json:"type,omitempty"`
```
[the remark this answers is not in the document]
Comment (lingdie) [reply controllers/image/api/v1/image_types.go:33]: Documents and descriptions need to be synchronized with the image name or with the image tag?

[the remark this answers is not in the document]
Comment (lingdie) [reply controllers/image/api/v1/image_types.go]: sample yaml
```yaml
apiVersion: image.sealos.io/v1
kind: Image
metadata:
  name: image-sample
spec:
  name: labring/mysql-operator
  url: www.google.com
  type: database
  tagIDMap:
    latest: mysql-latest-id
    8.0.23-14.1: mysql-8.0.23-14.1-id
  icon: https://icons8.com/icon/jYQqxG8eyJ19/external-mysql-programming-and-development-those-icons-lineal-color-those-icons
  description: mysql image description
  docs: msql docs
```

File: controllers/app/main.go (hunk 1 above)
Comment (lingdie) [reply controllers/app/main.go:74]: Done

File: controllers/image/api/v1/image_types.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	//+kubebuilder:validation:Required
+	Tags []string `json:"tags"`
+}
+
+type ImageDetailInfo struct {
+	URL         string         `json:"url,omitempty"`
+	Keywords    []ImageKeyword `json:"keywords,omitempty"`
+	Description string         `json:"description,omitempty"`
+	Icon        string         `json:"icon,omitempty"`
+	Docs        string         `json:"docs,omitempty"`
```
Comment (fanux) [inline controllers/image/api/v1/image_types.go:52]: What if every version has diff docs and description?

Comment (lingdie) [reply controllers/image/api/v1/image_types.go:52]: So Documents and descriptions need to be synchronized with the image name + image tag.
Refer to this [issue](https://github.com/labring/sealos/issues/1902).

File: controllers/app/api/v1/appaction_types.go [hunk 4]
```diff
@@ -23,48 +23,42 @@ import (
 // EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
-// ActionsSpec defines the desired state of Actions
-type ActionsSpec struct {
+// AppActionSpec defines the desired state of AppAction
+type AppActionSpec struct {
```
Comment (fanux) [inline controllers/app/api/v1/appaction_types.go:27]: I think the name `Action` is better then `AppAction` because Action maybe not only for app~

Comment (lingdie) [reply controllers/app/api/v1/appaction_types.go:27]: ok

File: controllers/app/api/v1/app_types.go [hunk 5]
```diff
@@ -17,41 +17,49 @@ limitations under the License.
...
-	Actions []string `json:"actions,omitempty"`
-	Workdir string   `json:"workdir,omitempty"`
-	Image   string   `json:"image,omitempty"`
+	// ImageInfo define image info that app need to use
+	//+kubebuilder:validation:Required
+	ImgInfo ImageInfo `json:"imgInfo"`
+
+	// Namespace is where app install in
+	// todo add requied and default namesapace as user's namespace
+	Namespace string `json:"namespace"`
```
Comment (fanux) [inline controllers/app/api/v1/app_types.go:51]: The metadata seems already has namespace

**Review by zzjin — APPROVED** [review]

**Pull request closed without merging**